File: programmers/[42577] 전화번호 목록/cpp/42577.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <string>

using namespace std;

class TrieNode {
public:
    unordered_map<char, TrieNode*> children;
    bool end;

    TrieNode() {
        end = false;
    }
};

class Trie {
public:
    TrieNode* root;

    Trie() {
        root = new TrieNode();
    }

    bool insert(const string& numbers) {
        TrieNode* curr = root;
        for (char number : numbers) {
            if (curr->children.find(number) == curr->children.end()) {
                curr->children[number] = new TrieNode();
            }
            if (curr->end == true) {
                return false;
            }
            curr = curr->children[number];
        }
        curr->end = true;
        return true;
    }
};


bool solution(vector<string> phone_book) {
    Trie phone_trie;  // 전화번호 Trie

    // 전화번호를 Trie에 집어넣기
    for (const string& phone_number : phone_book) {
        if (!phone_trie.insert(phone_number)) {
            return false;  // 접두어가 있다면 바로 False
        }
    }

    // 확인
    for (const string& phone_number : phone_book) {
        TrieNode* curr = phone_trie.root;
        for (char number : phone_number) {
            curr = curr->children[number];
        }
        if (!curr->children.empty()) {
            return false;  // 끝 숫자인데 아래 children이 있으면 접두어
        }
    }

    return true;  // 접두어가 없다면 True
}

File: programmers/[42577] 전화번호 목록/python/42577.py
from typing import *

class TrieNode:
    def __init__(self):
        self.children: dict = {}
        self.end: bool = False
    
class Trie:
    def __init__(self):
        self.root: TrieNode = TrieNode()
    
    # 삽입 과정에서 접두어가 있으면 False 반환
    def insert(self, numbers) -> bool:
        curr: TrieNode = self.root
        for number in numbers:
            if number not in curr.children:
                curr.children[number] = TrieNode()
            curr = curr.children[number]
            
            # 이미 다른 번호가 접두어로 존재하는 경우
            if curr.end:
                return False
        
        # 해당 번호가 다른 번호의 접두어인 경우
        if curr.children:
            return False
        
        # 번호 삽입 완료
        curr.end = True
        return True


def solution(phone_book: List[str]) -> bool: 
    phone_trie: Trie = Trie()     # 전화번호 Trie
    # 전화번호를 Trie에 집어넣기
    for phone_number in phone_book:
        if not phone_trie.insert(phone_number):
            return False
    
    return True
        
    
        
        
        